### PR TITLE
StateProofs: Use stateproof tracker

### PR DESCRIPTION
**Pull request by algoidan**

### CODE REVIEW

File: stateproof/verify/stateproof.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ package verify
 import (
 	"errors"
 	"fmt"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
```
Comment (almog-t) [inline stateproof/verify/stateproof.go:22]: nits: redundant newline

File: test/e2e-go/features/stateproofs/stateproofs_test.go [hunk 2]
```diff
@@ -382,7 +382,7 @@ func TestStateProofMessageCommitmentVerification(t *testing.T) {
 }
 
 func getDefaultStateProofConsensusParams() config.ConsensusParams {
-	consensusParams := config.Consensus[protocol.ConsensusCurrentVersion]
+	consensusParams := config.Consensus[protocol.ConsensusFuture]
```
Comment (almog-t) [inline test/e2e-go/features/stateproofs/stateproofs_test.go:385]: We should remember to change this after the consensus upgrade. Maybe add a ticket?

File: ledger/apply/stateproof.go [hunk 3]
```diff
@@ -41,31 +42,54 @@ func StateProof(tx transactions.StateProofTxnFields, atRound basics.Round, sp St
...
-		if err != nil {
-			return err
-		}
+	atRoundHdr, err := sp.BlockHdr(atRound)
+	if err != nil {
+		return err
+	}
+
+	var verificationData *ledgercore.StateProofVerificationData
+	if config.Consensus[atRoundHdr.CurrentProtocol].StateProofUseTrackerVerification {
```
Comment (almog-t) [inline ledger/apply/stateproof.go:57]: Do you think we should cover the verification method switch with a test?

**Review by almog-t — COMMENTED** [review]
Good work, not a lot we can or should change.

File: stateproof/verify/stateproof.go [hunk 4]
```diff
@@ -51,19 +50,24 @@ func AcceptableStateProofWeight(votersHdr *bookkeeping.BlockHeader, firstValid b
 	latestRoundInProof := votersHdr.Round + basics.Round(proto.StateProofInterval)
 	total := votersHdr.StateProofTracking[protocol.StateProofBasic].StateProofOnlineTotalWeight
 
+	return calculateAcceptableStateProofWeight(total, &proto, latestRoundInProof, firstValid, logger)
+}
+
+func calculateAcceptableStateProofWeight(total basics.MicroAlgos, proto *config.ConsensusParams, lastAttestedRound basics.Round, firstValid basics.Round, logger logging.Logger) uint64 {
```
[the remark this answers is not in the document]
Comment (algoidan) [reply stateproof/verify/stateproof.go:56]: This function and `AcceptableStateProofWeight` are being invoked not only on the verification flow but also when the proven checks if he has enough signatures. I think it is weird to use a verification structure when you want to evaluate the weight threshold.

File: ledger/apply/stateproof.go (hunk 3 above)
Comment (algoidan) [reply ledger/apply/stateproof.go:57]: There are two tests - 
1 TestCowStateProof check the old code - verification using blockheader
2 TestCowStateProofV34 - using the tracker.

**Review by almog-t — APPROVED** [review]

File: ledger/internal/eval_test.go [hunk 5]
```diff
@@ -266,13 +281,116 @@ func TestCowStateProof(t *testing.T) {
...
+	spType = protocol.StateProofType(1234) // bad stateproof type
+	stateProofTx := transactions.StateProofTxnFields{
+		StateProofType: spType,
+		StateProof:     stateProof,
+		Message:        msg,
+	}
+	err := apply.StateProof(stateProofTx, atRound, c0, validate)
+	require.ErrorIs(t, err, apply.ErrStateProofTypeNotSupported)
 
-	// 100% coverage
```
Comment (algorandskiy) [inline ledger/internal/eval_test.go:275]: no more 100% coverage?

**Review by algorandskiy — COMMENTED** [review]
Looks reasonable, and it looks like a time to review the entire change

File: ledger/internal/eval_test.go (hunk 5 above)
Comment (algoidan) [reply ledger/internal/eval_test.go:275]: just thought that the comments doesn't help much :)
If you think there is an edge case that we didn't cover, I will happily test it
